File: Matlab/calculus/find_root/use_func_bisection.m
clc;
clear;
close all;

% Input
a = 0;
b = 2;
e = 0.00001;
F = @(x)(x.^4/4 + x.^2/2 + x - 2);

% Algorithm
fprintf("I\ta\t\t\t\tf(a)\t\t\tb\t\t\t\tf(b)\t\t\tc\t\t\t\tf(c)\n");
for i = 1:100
    af = F(a);
    bf = F(b);
    NsolF = (a+b)/2;
    NvalF = F(NsolF);
    fprintf("%d)\t%.10f\t%.10f\t%.10f\t%.10f\t%.10f\t%.10f\n", i, a, af, b, bf, NsolF, NvalF);
    if abs(NvalF) < e
        break;
    elseif af*NvalF > 0
        a = NsolF;
    else
        b = NsolF;
    end
end

% Output
fprintf("\nNsolF = %.10f\n", NsolF);

% Compare
EsolF = fzero(F, (a+b)/2);
fprintf("EsolF = %.10f\nError = %.10f\n", EsolF, abs(EsolF - NsolF));

% **************************************************^**************************************************
% *****************************# Copyright by Ali Forouzandeh Hafshejani #*****************************
% ******************************# Scientific Computing Specialist 20@21 #******************************
% **************************************************^**************************************************

File: C++/calculus/find_root/use/use_func_bisection.cpp
// Start
#include <complex>
#include <iomanip>
#include <iostream>
using namespace std;

// Function
double func(double x)
{
    return (
        pow(x, 4)/4 + pow(x, 2)/2 + x - 2
    );
}

// Main
int main()
{
    int iter, Niter;
    double ax, bx, err, af, bf, NsolF, NvalF;

    // Input
    ax = 0;
    bx = 2;
    err = 0.00001;
    Niter = 100;

    // Algorithm
    cout << "I\ta\t\tf(a)"
        << "\t\tb\t\tf(b)"
        << "\t\tc\t\tf(c)"
        << setprecision(10) << fixed
        << endl;

    for (iter = 0; iter < Niter; ++iter)
    {
        af = func(ax);
        bf = func(bx);
        NsolF = (ax + bx)/2;
        NvalF = func(NsolF);

        cout << iter+1 << ")\t"
            << ax << '\t' << af << '\t'
            << bx << '\t' << bf << '\t'
            << NsolF << '\t' << NvalF << endl;

        if (abs(NvalF) < err)
        {
            break;
        }
        else if (af*NvalF > 0)
        {
            ax = NsolF;
        }
        else
        {
            bx = NsolF;
        }
    }

    // Output
    cout << endl << "NsolF = " << NsolF << endl;

    // End
    return 0;
}

// **************************************************^**************************************************
// *****************************# Copyright by Ali Forouzandeh Hafshejani #*****************************
// ******************************# Scientific Computing Specialist 20@21 #******************************
// **************************************************^**************************************************